lib/util/glob-util.js

```diff
@@ -160,10 +154,25 @@ function listFilesToProcess(globPatterns, options) {
         const file = path.resolve(cwd, pattern);
 
         if (shell.test("-f", file)) {
-            addFile(fs.realpathSync(file), !shell.test("-d", file));
+            const ignoredPaths = new IgnoredPaths(options);
+
+            addFile(fs.realpathSync(file), !shell.test("-d", file), ignoredPaths);
         } else {
+
+            // regex to find .hidden or /.hidden patterns, but not ./relative or ../relative
+            const globIncludesDotfiles = /(?:(?:^\.)|(?:[\/\\]\.))[^\/\\\.].*/.test(pattern);
+
+            options.dotfiles = options.dotfiles || globIncludesDotfiles;
```

Comment:
Dotfiles/dotfolders will not be linted in most cases, only if the user explicitly does one of the following:

1. specifies a glob which would include dotfiles (e.g. `.*` or `test/**/.*`) (this is what the regex is to find)
2. unignores dotfiles with an `--ignore-pattern` or pattern in an `.eslintignore` file.  This is similar to our other default ignores of `node_modules` and `bower_components`